### PR TITLE
Update explorer to re-use tree nodes instead of creating new nodes on each render

### PR DESCRIPTION
**Pull request by mpiroc**
Currently, the explorer creates new tree nodes on each render. With this change, nodes are re-used where applicable. This allows us much more fine-grained control over updates.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [x] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [x] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

* Each node now contains:
  * a map of its children, which is updated when `getChildren` is invoked.
  * the code for the node's associated region (where applicable)
* New utility functions allow us to efficiently update the children map without recreating nodes.
* New wrapper types around AWS SDK service clients allow them to be mocked.
* We now build against the same NodeJS version that vscode is built against.
  * This allows us to use async generator functions.
  * In the new version, `assert.equals` is deprecated in favor of `assert.strictEquals`.

## Motivation and Context

The current design of the explorer creates new instances of nodes every time that `getChildren` is called. This has numerous disadvantages (for example, due to the structure of the vscode API, we can't refresh a node without a reference to its instance, which forces us to refresh the whole tree) and hamstrings us for future changes.

This change refactors the explorer to re-use nodes whenever possible, and hold on to their references. This allows us to quickly update a granular portion of the tree. For example, after a node is deleted, we refresh its parent. Previously, we refreshed the entire tree.

## Related Issue(s)

#175

## Testing

* Updated and ran all existing unit tests
* Added new unit tests for collection utility methods.

More test coverage is incoming, but I wanted to get some eyeballs on this sooner rather than later.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
  - In progress
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/shared/awsClientBuilder.test.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import * as assert from 'assert'
 import { Service } from 'aws-sdk'
 import { ServiceConfigurationOptions } from 'aws-sdk/lib/service'
 import * as vscode from 'vscode'
-import { AWSClientBuilder } from '../../shared/awsClientBuilder'
+import { DefaultAWSClientBuilder } from '../../shared/awsClientBuilder'
 import { AwsContext, ContextChangeEventsArgs } from '../../shared/awsContext'
 
 describe('AwsClientBuilder', () => {
```
Comment (awschristou) [inline src/test/shared/awsClientBuilder.test.ts]: should this be describe `DefaultAWSClientBuilder` now instead? If so, consider renaming the file too.

File: src/lambda/commands/deleteLambda.ts [hunk 2]
```diff
@@ -9,29 +9,28 @@ import * as nls from 'vscode-nls'
...
-import { RegionFunctionNode } from '../explorer/functionNode'
+import { StandaloneFunctionNode } from '../explorer/standaloneNodes'
 
-export async function deleteLambda(node: RegionFunctionNode, refresh: () => void) {
-    if (!node || !node.info.configuration.FunctionName) {
+export async function deleteLambda(
+    node: StandaloneFunctionNode,
+    refresh: () => void
+) {
+    const client = ext.toolkitClientBuilder.createLambdaClient(node.regionCode)
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/lambda/commands/deleteLambda.ts]: (Comment applies everywhere createXXXClient is called)

File: src/lambda/commands/deleteLambda.ts [hunk 3]
```diff
@@ -9,29 +9,28 @@ import * as nls from 'vscode-nls'
...
+
+    if (!node || !node.configuration.FunctionName) {
```
[the remark this answers is not in the document]
Comment (mpiroc) [reply src/lambda/commands/deleteLambda.ts]: ✔️

File: src/lambda/commands/deleteLambda.ts [hunk 4]
```diff
@@ -9,29 +9,28 @@ import * as nls from 'vscode-nls'
...
 import { ext } from '../../shared/extensionGlobals'
-import { RegionFunctionNode } from '../explorer/functionNode'
+import { StandaloneFunctionNode } from '../explorer/standaloneNodes'
 
-export async function deleteLambda(node: RegionFunctionNode, refresh: () => void) {
-    if (!node || !node.info.configuration.FunctionName) {
+export async function deleteLambda(
+    node: StandaloneFunctionNode,
+    refresh: () => void
+) {
```
[the remark this answers is not in the document]
Comment (mpiroc) [reply src/lambda/commands/deleteLambda.ts]: ✔️

File: src/lambda/commands/deleteLambda.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpiroc) [reply src/lambda/commands/deleteLambda.ts]: ✔️

File: src/lambda/explorer/cloudFormationNodes.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import * as vscode from 'vscode'
+import { ext } from '../../shared/extensionGlobals'
+import { AWSTreeNodeBase } from '../../shared/treeview/awsTreeNodeBase'
+import { intersection, toArrayAsync, toMap, toMapAsync, updateInPlace } from '../collectionUtils'
+import { listCloudFormationStacks, listLambdaFunctions } from '../utils'
+import { FunctionNodeBase } from './functionNode'
+import { PlaceholderNode } from './placeholderNode'
+import { RegionNode } from './regionNode'
+
+export class CloudFormationStackGroupNode extends AWSTreeNodeBase {
```
[the remark this answers is not in the document]
Comment (mpiroc) [reply src/lambda/explorer/cloudFormationNodes.ts]: ✔️

**Pull request merged**